library/network/src/modules/SuSEFirewall.rb

```diff
@@ -403,6 +405,255 @@ def _sf2_to_firewalld_service(service)
+            to_add = good_ports - current_ports
+            Builtins.y2debug("Ports: drop -> #{to_drop} add -> #{to_add}")
+            to_drop.each { |rmp| @fwd_api.remove_port_from_zone(zone, rmp) }
+            to_add.each { |gp| @fwd_api.add_port_to_zone(zone, gp) }
+            @SETTINGS[zone][:modified].delete(:ports)
+          end
+
+          if @SETTINGS[zone][:modified].include?(:protocols)
+            # These are the protocols which should be enabled
+            good_protocols = @SETTINGS[zone][:protocols]
+            # These are the ones which are enabled
+            current_protocols = @fwd_api.get_protocols_in_zone(zone)
+            # And these are the ones which should be dropped
+            to_drop = current_protocols - good_protocols
+            to_add = good_protocols - current_protocols
+            Builtins.y2debug("Protocols: drop -> #{to_drop} add -> #{to_add}")
+            to_drop.each { |rmp| @fwd_api.remove_protocol_from_zone(zone, rmp) }
+            to_add.each { |gp| @fwd_api.add_protocol_to_zone(zone, gp) }
+            @SETTINGS[zone][:modified].delete(:protocols)
+          end
+
+          # Configuration is now live. Move on
+          ResetModified()
+        end
+
+      rescue FirewallCMDError
+        Builtins.y2error("firewall-cmd failed")
+        raise
+      end
+
+      # FIXME: perhaps "== true" can be dropped since this should
+      # always be boolean?
+      if !@SETTINGS["enable_firewall"].nil?
+        if @SETTINGS["enable_firewall"] == true
+          Builtins.y2milestone("Enabling firewall services")
+          return false if !EnableServices()
+        else
+          Builtins.y2milestone("Disabling firewall services")
+          return false if !DisableServices()
+        end
+      end
+
+      true
+    end
+
+
+    # In SF2, it's used to write configuration, but not activate. For firewalld
+    # this is simply here to satisfy callers, like modules/Nfs.rb.
+    # @return true
+    def WriteOnly
+      # This does not check if firewalld is running
+      return false if !WriteConfiguration()
+    end
+
+    # Function which starts/stops firewall. Then firewall is started immediately
+    # when firewall is wanted to be started: SetStartService(boolean). FirewallD
+    # needs to be reloaded instead of doing a full-blown restart to get the new
+    # configuration up and running.
+    #
+    # @return	[Boolean] if successful
+    def ActivateConfiguration
+      # starting firewall during second stage can cause deadlock in systemd - bnc#798620
+      # Moreover, it is not needed. Firewall gets started via dependency on multi-user.target
+      # when second stage is over.
+      if Mode.installation
+        Builtins.y2milestone("Do not touch firewall services during installation")
+
+        return true
+      end
+
+      # We need a running service if we want to reload the permanent
+      # configuration
+      unless IsStarted()
+        StartServices()
```

Comment:
You are right. I can't really say I remember why I did it this way. I will change it.